File: p532m/find_pairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findPairs = function (nums, k) {
  const count = new Map();
  for (let n of nums) {
    let c = count.get(n) || 0;
    count.set(n, c + 1);
  }
  let ans = 0;
  count.forEach((v, n) => {
    if ((k === 0 && count.get(n) > 1) || (k > 0 && count.has(n + k))) ans++;
  });
  return ans;
};
// TESTS
[
  {
    nums: [3, 1, 4, 1, 5],
    k: 2,
    expected: 2,
  },
  {
    nums: [1, 2, 3, 4, 5],
    k: 1,
    expected: 4,
  },
  {
    nums: [1, 3, 1, 5, 4],
    k: 0,
    expected: 1,
  },
  {
    nums: [1, 2, 4, 4, 3, 3, 0, 9, 2, 3],
    k: 3,
    expected: 2,
  },
  {
    nums: [-1, -2, -3],
    k: 1,
    expected: 2,
  },
].forEach(({ nums, k, expected }) => {
  const actual = findPairs(nums, k);
  console.log(`The # of ${k}-diff pairs in array`, nums, '->', actual);
  console.assert(actual === expected);
});

File: p532m/find_pairs.py
from collections import Counter


class Solution:
    def findPairs(self, nums: list[int], k: int) -> int:
        ans, count = 0, Counter(nums)
        for n in count:
            if k == 0 and count[n] > 1 or k > 0 and n + k in count:
                ans += 1
        return ans


# TESTS
for nums, k, expected in [
    ([3, 1, 4, 1, 5], 2, 2),
    ([1, 2, 3, 4, 5], 1, 4),
    ([1, 3, 1, 5, 4], 0, 1),
    ([1, 2, 4, 4, 3, 3, 0, 9, 2, 3], 3, 2),
    ([-1, -2, -3], 1, 2),
]:
    sol = Solution()
    actual = sol.findPairs(nums, k)
    print(f"The # of {k}-diff pairs in array", nums, "->", actual)
    assert actual == expected
